File: python/166_FractiontoRecurringDecimal.py
from unittest import TestCase
# https://leetcode.com/problems/fraction-to-recurring-decimal/


class FractiontoRecurringDecimal(TestCase):
    def fractionToDecimal(self, numerator, denominator):
        """
        :type numerator: int
        :type denominator: int
        :rtype: str
        """
        if not numerator:
            return "0"
        flag = 1
        if numerator < 0:
            numerator *= -1
            flag *= -1
        if denominator < 0:
            denominator *= -1
            flag *= -1
        neg = "-" if flag == -1 else ""
        integer = int(numerator/denominator)
        remain = numerator - integer*denominator
        if not remain:
            return f"{neg}{integer}"

        occur = {}
        decimals = ""
        while remain and remain not in occur:
            occur[remain] = len(decimals)
            remain *= 10
            mul = int(remain/denominator)
            decimals += str(mul)
            remain -= mul*denominator

        if remain:
            return f"{neg}{integer}.{decimals[:occur[remain]]}({decimals[occur[remain]:]})"
        else:
            return f"{neg}{integer}.{decimals}"

    def test1(self):
        self.assertEqual("0.5", self.fractionToDecimal(1, 2))

    def test2(self):
        self.assertEqual("2", self.fractionToDecimal(2, 1))

    def test3(self):
        self.assertEqual("0.(6)", self.fractionToDecimal(2, 3))

    def test4(self):
        self.assertEqual("0.(012)", self.fractionToDecimal(4, 333))

    def test5(self):
        self.assertEqual("-6.25", self.fractionToDecimal(-50, 8))

    def test6(self):
        self.assertEqual("-2147483648", self.fractionToDecimal(-2147483648, 1))

    def test6(self):
        self.assertEqual("0", self.fractionToDecimal(0, -1))

File: kotlin/src/main/kotlin/166.FractiontoRecurringDecimal.kt
import org.junit.Assert.assertEquals
import org.junit.Test

//https://leetcode.com/problems/fraction-to-recurring-decimal/

class FractiontoRecurringDecimal{
  fun fractionToDecimal(numerator: Int, denominator: Int): String {
    if(numerator == 0) return "0"

    var flag = 1
    var num = numerator.toLong()
    var den = denominator.toLong()
    if(num < 0) {
      flag *= -1
      num *= -1
    }
    if(den < 0) {
      flag *= -1
      den *= -1
    }

    val neg = if(flag == -1) "-" else ""
    val int = num/den
    var remain = num - int*den
    if(remain == 0L) return "${neg}${int}"

    val occur = HashMap<Long, Int>()
    val decimals = buildString {
      while(remain != 0L && !occur.containsKey(remain)){
        occur[remain] = length
        remain *= 10
        val mul = remain/den
        append(mul)
        remain -= mul*den
      }
    }

    return if(remain == 0L)
      "${neg}${int}.${decimals}"
    else
      "${neg}${int}.${decimals.substring(0, occur[remain]!!)}(${decimals.substring(occur[remain]!!)})"
  }

  @Test
  fun test1(){
    assertEquals("0.5", fractionToDecimal(1, 2))
  }

  @Test
  fun test2(){
    assertEquals("2", fractionToDecimal(2, 1))
  }

  @Test
  fun test3(){
    assertEquals("0.(6)", fractionToDecimal(2, 3))
  }

  @Test
  fun test4(){
    assertEquals("0.(012)", fractionToDecimal(4, 333))
  }

  @Test
  fun test5(){
    assertEquals("-2147483648", fractionToDecimal(-2147483648, 1))
  }
}